crates/matrix-sdk/src/oidc/auth_code_builder.rs

```diff
@@ -0,0 +1,248 @@
+/// with [`Self::build()`].
+#[allow(missing_debug_implementations)]
+pub struct OidcAuthCodeUrlBuilder {
+    oidc: Oidc,
+    scope: Scope,
+    redirect_uri: Url,
+    display: Option<Display>,
+    prompt: Option<Vec<Prompt>>,
+    max_age: Option<NonZeroU32>,
+    ui_locales: Option<Vec<LanguageTag>>,
+    login_hint: Option<String>,
+    acr_values: Option<HashSet<String>>,
+}
+
+impl OidcAuthCodeUrlBuilder {
+    pub(super) fn new(oidc: Oidc, scope: Scope, redirect_uri: Url) -> Self {
+        Self {
+            oidc,
+            scope,
+            redirect_uri,
+            display: None,
+            prompt: None,
+            max_age: None,
+            ui_locales: None,
+            login_hint: None,
+            acr_values: None,
+        }
+    }
+
+    /// Set how the Authorization Server should display the authentication and
+    /// consent user interface pages to the End-User.
+    pub fn display(mut self, display: Display) -> Self {
+        self.display = Some(display);
+        self
+    }
+
+    /// Set the [`Prompt`] of the authorization URL.
+    ///
+    /// [`Prompt::Create`] can be used to signify that the user wants to
+    /// register a new account. If [`Prompt::None`] is used, it must be the only
+    /// value.
+    pub fn prompt(mut self, prompt: Vec<Prompt>) -> Self {
+        self.prompt = Some(prompt);
+        self
+    }
+
+    /// Set the allowable elapsed time in seconds since the last time the
+    /// End-User was actively authenticated by the OpenID Provider.
+    pub fn max_age(mut self, max_age: NonZeroU32) -> Self {
+        self.max_age = Some(max_age);
+        self
+    }
+
+    /// Set the preferred locales of the user.
+    ///
+    /// Must be ordered from the preferred locale to the least preferred locale.
+    pub fn ui_locales(mut self, ui_locales: Vec<LanguageTag>) -> Self {
+        self.ui_locales = Some(ui_locales);
+        self
+    }
+
+    /// Set the hint to the Authorization Server about the login identifier the
+    /// End-User might use to log in.
+    ///
+    /// To set a Matrix user ID as a login hint, use [`Self::user_id_hint()`].
+    ///
+    /// Erases any value set with [`Self::user_id_hint()`].
+    pub fn login_hint(mut self, login_hint: String) -> Self {
+        self.login_hint = Some(login_hint);
+        self
+    }
+
+    /// Set the hint to the Authorization Server about the Matrix user ID the
+    /// End-User might use to log in.
+    ///
+    /// To set another type of identifier as a login hint, use
+    /// [`Self::login_hint()`].
+    ///
+    /// Erases any value set with [`Self::login_hint()`].
+    pub fn user_id_hint(mut self, user_id: &UserId) -> Self {
+        self.login_hint = Some(format!("mxid:{user_id}"));
+        self
+    }
+
+    /// Set the requested Authentication Context Class Reference values.
+    ///
+    /// This is only necessary with specific providers.
+    pub fn acr_values(mut self, acr_values: HashSet<String>) -> Self {
```

Comment:
If it's "Authentication Context Class Reference", why is the abbreviation `acr` instead of `accr`?

Reply:
I don't know, but that's the abbreviation used by OIDC. Also this is the exact name of the field this is supposed to fill.